src/wallet/spend.cpp

```diff
@@ -609,23 +618,77 @@ static uint32_t GetLocktimeForNewTransaction(interfaces::Chain& chain, const uin
     // enough, that fee sniping isn't a problem yet, but by implementing a fix
     // now we ensure code won't be written that makes assumptions about
     // nLockTime that preclude a fix later.
+    //
+    // nSequence-based
+    // ---------------
+    //
+    // If all inputs in the tx spend taproot coins, use nSequence to discourage
+    // fee-sniping. This creates a "privacy cloak" for txs that use taproot
+    // key-path spends (BIP326).
+    //
+    // Compared to nLockTime-based anti-fee-sniping, this may be minimally less
+    // effective, as nSequence only allows relative locks. If all inputs of the
+    // tx can be moved backward in the chain, this tx can be as well. Though,
+    // this should be fine given that 10% of txs have their locktime randomly
+    // reduced to a past block. Also, sequence-based locking will only be used
+    // if every sequence is at least 1.
     if (IsCurrentForAntiFeeSniping(chain, block_hash)) {
-        locktime = block_height;
-
-        // Secondly occasionally randomly pick a nLockTime even further back, so
-        // that transactions that are delayed after signing for whatever reason,
-        // e.g. high-latency mix networks and some CoinJoin implementations, have
-        // better privacy.
-        if (GetRandInt(10) == 0)
-            locktime = std::max(0, (int)locktime - GetRandInt(100));
+        bool locktime_based{false};
+        for (const auto& in : inputs) {
+            std::vector<std::vector<uint8_t>> dummy;
+            const TxoutType in_type{Solver(in.txout.scriptPubKey, dummy)};
+            // Use locktime if any nSequence is out-of-range or any input isn't taproot
+            if (1 > in.m_depth || in.m_depth > 65535 || in_type != TxoutType::WITNESS_V1_TAPROOT) {
+                locktime_based = true;
+                break;
+            }
+        }
+        if (locktime_based || GetRandInt(2) == 0) {
+            tx.nLockTime = block_height;
+
+            // Secondly occasionally randomly pick a nLockTime even further back, so
+            // that transactions that are delayed after signing for whatever reason,
+            // e.g. high-latency mix networks and some CoinJoin implementations, have
+            // better privacy.
+            if (GetRandInt(10) == 0) {
+                tx.nLockTime = std::max(0, (int)tx.nLockTime - GetRandInt(100));
```

Comment:
please convert this to a c++11 style functional cast

Reply:
Seems unrelated to this pull request, but done in https://github.com/bitcoin/bitcoin/pull/24225